src/array.rs

```diff
@@ -506,26 +521,20 @@ impl<T: TypeNum, D: Dimension> PyArray<T, D> {
     /// Get the immutable view of the internal data of `PyArray`, as
     /// [`ndarray::ArrayView`](https://docs.rs/ndarray/latest/ndarray/type.ArrayView.html).
     ///
-    /// # Example
-    /// ```
-    /// # #[macro_use] extern crate ndarray;
-    /// use numpy::PyArray;
-    /// let gil = pyo3::Python::acquire_gil();
-    /// let py_array = PyArray::arange(gil.python(), 0, 4, 1).reshape([2, 2]).unwrap();
-    /// assert_eq!(
-    ///     py_array.as_array(),
-    ///     array![[0, 1], [2, 3]]
-    /// )
-    /// ```
-    pub fn as_array(&self) -> ArrayView<'_, T, D> {
-        self.type_check_assert();
-        unsafe { ArrayView::from_shape_ptr(self.ndarray_shape(), self.data()) }
+    /// # Safety
+    /// If the internal array is not readonly and can be mutated from Python code,
+    /// holding the `ArrayView` might cause an undefined undefined behavior.
+    pub unsafe fn as_array(&self) -> ArrayView<'_, T, D> {
+        ArrayView::from_shape_ptr(self.ndarray_shape(), self.data())
     }
 
-    /// Almost same as [`as_array`](#method.as_array), but returns `ArrayViewMut`.
-    pub fn as_array_mut(&self) -> ArrayViewMut<'_, T, D> {
-        self.type_check_assert();
-        unsafe { ArrayViewMut::from_shape_ptr(self.ndarray_shape(), self.data()) }
+    /// Returns the internal array as `ArrayViewMut`. See also [`as_array`](#method.as_array).
+    ///
+    /// # Safety
+    /// If another reference to the internal data exists(e.g., `&[T]` or `ArrayView`),
+    /// it might cause an undefined undefined behavior.
+    pub unsafe fn as_array_mut(&self) -> ArrayViewMut<'_, T, D> {
```

Comment:
If renaming to `as_mut_slice`, might want to consider this being `as_mut_array`, but it's less clear cut.